### PR TITLE
Datagrid Adapter: TreeListAdapter

### PR DESCRIPTION
**Pull request by alexander-schranz**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | yes
| BC breaks? | no
| Deprecations? | no
| License | MIT
| Documentation PR | sulu/sulu-docs#prnum

#### What's in this PR?

Tree component structure.

#### Why?

Needed for tree datagrid component.

#### To Do

- [x] Finish PR
- [x] Tests

### CODE REVIEW

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Body.js [hunk 1]
```diff
@@ -11,12 +11,17 @@ type Props = {
     /** @ignore */
     selectMode?: SelectMode,
     /** @ignore */
+    selectInFirstCell?: boolean,
+    /** @ignore */
     onRowSelectionChange?: (rowId: string | number, selected?: boolean) => void,
+    /** @ignore */
+    onRowToggleChange?: (rowId: string | number, expanded: boolean) => void,
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Body.js]: Why do we need a new callback?

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Header.js [hunk 2]
```diff
@@ -63,17 +68,41 @@ export default class Header extends React.PureComponent<Props> {
...
+        return <div className={tableStyles.cellContent}>
+            <div className={tableStyles.cellSelect}>
+                <Checkbox
+                    skin="light"
+                    checked={!!this.props.allSelected}
+                    onChange={this.handleAllSelectionChange}
+                />
+            </div>
+            {children}
+        </div>;
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Header.js]: Multiline JSX must be wrapped in brackets.

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Row.js [hunk 3]
```diff
@@ -64,55 +86,97 @@ export default class Row extends React.PureComponent<Props> {
...
+        let toggler = null;
+        let select = null;
+        let style = {};
+
+        if (this.props.hasChildren) {
+            toggler = this.createToggler();
+        }
+
+        if (this.props.depth) {
+            style.paddingLeft = (this.props.depth * 25) + 'px'
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Row.js]: No magic numbers

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Row.js [hunk 4]
```diff
@@ -140,6 +204,15 @@ export default class Row extends React.PureComponent<Props> {
         });
     };
 
+    handleToggleChange = () => {
+        const {id, rowIndex, expanded} = this.props;
+        const identifier = id || rowIndex;
+
+        if (this.props.onToggleChange && identifier) {
+            this.props.onToggleChange(identifier, !expanded);
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Row.js]: Now I see what this callback is for :see_no_evil: I think `toggle` is a little bit an unlucky way, because you can toggle any flag. Maybe we should replace it by two callbacks: `onExpand` and `onCollapse`? Then it would be clearer what it is doing.

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/table.scss [hunk 5]
```diff
@@ -70,6 +70,21 @@ $rowHeight: 40px;
...
+}
+
+.toggle-icon {
+    cursor: pointer;
+    display: inline-block;
+    width: 20px;
+}
+
+.cell-select {
+    padding-right: 11px;
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/table.scss]: Normally we try to use rounded values.

File: src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/Datagrid.js [hunk 6]
```diff
@@ -72,8 +72,7 @@ export default class Datagrid extends React.Component<Props> {
 
     handleItemSelectionChange = (id: string | number, selected?: boolean) => {
         const {store} = this.props;
-        // TODO do not hardcode id but use metdata instead
-        const row = store.data.find((item) => item.id === id);
+        const row = store.find(id);
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/Datagrid.js]: Maybe `findById`?

File: src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/adapters/TreeListAdapter.js [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+        dataList = this.flattenData(this.props.data, dataList, depth);
+
+        return dataList;
+    }
+
+    isExpanded(identifier: string | number) {
+        return this.expandedRows.includes(identifier);
+    }
+
+    flattenData(items: Array<Object>, dataList: Array<Object>, depth: number) {
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/adapters/TreeListAdapter.js]: My gut feeling is that it would be easier to simply loop over the items in the `render` function instead of building another object.

File: src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/adapters/TreeListAdapter.js [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+
+    renderRows(items: Array<Object>) {
+        return items.map((item) => {
+            const {
+                schema,
+                selections,
+            } = this.props;
+            const {data} = item;
+            const schemaKeys = Object.keys(schema);
+            let loading = false;
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/adapters/TreeListAdapter.js]: let not necessary here if we omit the if and assign the returned boolean directly.

File: src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/structureStrategies/TreeStructureStrategy.js [hunk 9]
```diff
@@ -28,6 +28,24 @@ export default class TreeStructureStrategy implements StructureStrategyInterface
         return this.findChildrenForParentId(this.data, parent);
     }
 
+    find(id: string | number): ?Object {
+        return this.findRecursive(this.data, id);
+    }
+
+    findRecursive(items: Array<Object>, id: string | number): ?Object {
+        for (let item of items) {
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/structureStrategies/TreeStructureStrategy.js]: I think `const` could be used here

File: src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/structureStrategies/TreeStructureStrategy.js [hunk 10]
```diff
@@ -28,6 +28,24 @@ export default class TreeStructureStrategy implements StructureStrategyInterface
...
+            if (item.data.id === id) {
+                return item.data;
+            }
+
+            let data = this.findRecursive(item.children, id);
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/structureStrategies/TreeStructureStrategy.js]: Could also be `const`

File: src/Sulu/Bundle/AdminBundle/Resources/js/index.js [hunk 11]
```diff
@@ -12,7 +12,7 @@ import Application from './containers/Application';
 import {fieldRegistry, Input, ResourceLocator, SingleSelect} from './containers/Form';
 import FieldBlocks from './containers/FieldBlocks';
 import {viewRegistry} from './containers/ViewRenderer';
-import {datagridAdapterRegistry, ColumnListAdapter, FolderAdapter, TableAdapter} from './containers/Datagrid';
+import {datagridAdapterRegistry, ColumnListAdapter, TreeListAdapter, FolderAdapter, TableAdapter} from './containers/Datagrid';
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/index.js]: Alphabet :)

File: src/Sulu/Bundle/ContentBundle/Resources/js/views/WebspaceOverview/WebspaceOverview.js [hunk 12]
```diff
@@ -142,7 +142,7 @@ class WebspaceOverview extends React.Component<ViewProps> {
                         </div>
                         <Datagrid
                             store={this.datagridStore}
-                            adapters={['column_list']}
+                            adapters={['tree_list', 'table', 'column_list']} // TODO
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Resources/js/views/WebspaceOverview/WebspaceOverview.js]: If we omit `table` it would not even be a todo :wink:

**Comment by danrot** [comment]
@alexander-schranz Do you plan on finishing that soon? That's something I could also need in #3855 :smile:

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Body.js (hunk 1 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Body.js]: this is needed for toggling. to open and close the children. for the new arrow in the Row.js (onToggleChange)

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Row.js (hunk 4 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Row.js]: @danrot sound good for me.

File: src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/adapters/TreeListAdapter.js (hunk 7 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/adapters/TreeListAdapter.js]: Did try that first but that didn't work as the Body.js not longer could set the callback function correctly. Try that with @trickreich at the hackday and first flatten the data did seems the only solution which did work then.

Or do you have a way where I can simple `array.push` the render elements and then render them as children from the Body.js.

**Comment by alexander-schranz** [comment]
@danrot own PR for the datagrid changes created #3860

**Comment by trickreich** [comment]
@danrot I'm finished on this

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Row.js [hunk 13]
```diff
@@ -64,61 +99,111 @@ export default class Row extends React.PureComponent<Props> {
...
+                onChange={this.handleMultipleSelectionChange}
+            />
         );
     };
 
     createButtonCells = () => {
-        const {id, rowIndex} = this.props;
-        const {buttons} = this.props;
+        const {buttons, rowIndex} = this.props;
+        const identifier = this.getIdentifier();
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Row.js]: With the above mentioned improvement we could directly use `this.identifier` in the JSX.

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Table.js [hunk 14]
```diff
@@ -19,11 +19,23 @@ type Props = {
     buttons?: Array<ButtonConfig>,
     /** Can be set to "single" or "multiple". Defaults is "none". */
     selectMode?: SelectMode,
+    /** Can be set to  */
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Table.js]: This comment is not finished, and I wouldn't add a comment here as well, as it doesn't add any value.

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/tests/__snapshots__/Table.test.js.snap [hunk 15]
```diff
@@ -423,6 +439,146 @@ exports[`Render the Table component in single selection mode 1`] = `
...
           </div>
         </td>
+        <td
+          class="cell"
+        >
+          <div
+            class="cellContent"
+          >
+            <div
+              class="cellContent"
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/tests/__snapshots__/Table.test.js.snap]: Why are there two nested `cellContent` divs?

File: src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/adapters/AbstractTableAdapter.js [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+import type {Schema} from '../types';
+import AbstractAdapter from './AbstractAdapter';
+
+export default class AbstractTableAdapter extends AbstractAdapter {
+    static defaultProps = {
+        data: [],
+    };
+
+    @computed get data(): Array<Object> {
+        return this.props.data;
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/adapters/AbstractTableAdapter.js]: Why can't we simply use the `this.props.data` observable instead of creating a separate computed value?

File: src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/adapters/TreeListAdapter.js [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+import FullLoadingStrategy from '../loadingStrategies/FullLoadingStrategy';
+import AbstractTableAdapter from './AbstractTableAdapter';
+
+@observer
+export default class TreeListAdapter extends AbstractTableAdapter {
+    static LoadingStrategy = FullLoadingStrategy;
+
+    static StructureStrategy = TreeStructureStrategy;
+
+    static icon = 'fa-sitemap';
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/adapters/TreeListAdapter.js]: Was that clarified with a designer?

File: src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/tests/adapters/TreeListAdapter.test.js [hunk 18]
```diff
@@ -0,0 +1,634 @@
...
+            page={1}
+            pageCount={1}
+            onItemClick={rowEditClickSpy}
+        />
+    );
+
+    expect(treeListAdapter).toMatchSnapshot();
+});
+
+test('Render data with plus button when onItemEdit callback is passed', () => {
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/tests/adapters/TreeListAdapter.test.js]: `when onItemAdd callback is passed`

File: src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/tests/adapters/TreeListAdapter.test.js [hunk 19]
```diff
@@ -0,0 +1,634 @@
...
+        <TreeListAdapter
+            data={data}
+            disabledIds={[]}
+            loading={false}
+            schema={schema}
+            selections={[]}
+            onPageChange={jest.fn()}
+            page={1}
+            pageCount={1}
+            onItemClick={rowEditClickSpy}
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/tests/adapters/TreeListAdapter.test.js]: The add button should be only added if the `onItemAdd` callback is passed, shouldn't it?

File: src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/tests/adapters/TreeListAdapter.test.js [hunk 20]
```diff
@@ -0,0 +1,634 @@
...
+        />
+    );
+
+    expect(treeListAdapter).toMatchSnapshot();
+});
+
+test('Render data with plus button when onItemEdit callback is passed', () => {
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/tests/adapters/TreeListAdapter.test.js]: Isn't that the same test as above?

File: src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/tests/adapters/__snapshots__/TableAdapter.test.js.snap [hunk 21]
```diff
@@ -58,7 +62,11 @@ exports[`Render data with all different visibility types schema 1`] = `
             <div
               class="cellContent"
             >
-              Description 2
+              <div
+                class="cellContent"
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/tests/adapters/__snapshots__/TableAdapter.test.js.snap]: Same as in above snapshot

File: src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/tests/adapters/__snapshots__/TreeListAdapter.test.js.snap [hunk 22]
```diff
@@ -0,0 +1,1242 @@
...
+        class="row"
+      >
+        <td
+          class="cell"
+        >
+          <div
+            class="cellContent"
+          >
+            <div
+              class="cellContent"
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/tests/adapters/__snapshots__/TreeListAdapter.test.js.snap]: Same as in above snapshot

File: src/Sulu/Bundle/AdminBundle/Resources/js/views/Datagrid/tests/__snapshots__/Datagrid.test.js.snap [hunk 23]
```diff
@@ -82,7 +82,11 @@ exports[`Should render the datagrid with a title 1`] = `
                 <div
                   class="cellContent"
                 >
-                  Description 1
+                  <div
+                    class="cellContent"
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/views/Datagrid/tests/__snapshots__/Datagrid.test.js.snap]: Same as above

File: src/Sulu/Bundle/ContentBundle/Resources/js/views/WebspaceOverview/WebspaceOverview.js [hunk 24]
```diff
@@ -138,11 +138,11 @@ class WebspaceOverview extends React.Component<ViewProps> {
                             </WebspaceSelect>
                         </div>
                         <Datagrid
-                            adapters={['column_list']}
-                            onAddClick={this.handleAddClick}
+                            store={this.datagridStore}
+                            adapters={['tree_list', 'column_list']}
                             onItemClick={this.handleEditClick}
+                            onAddClick={this.handleAddClick}
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Resources/js/views/WebspaceOverview/WebspaceOverview.js]: Why was this changed to a non-alphabethic order?

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Row.js [hunk 25]
```diff
@@ -18,16 +20,43 @@ type Props = {
...
+        selectInFirstCell: false,
+        hasChildren: false,
+        expanded: false,
+        isLoading: false,
         rowIndex: 0,
+        depth: 0,
+        depthPadding: 25,
+    };
+
+    getIdentifier = (): (string | number) => {
```
[the remark this answers is not in the document]
Comment (trickreich) [reply src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Row.js]: This component isn't a observer at all

File: src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/adapters/TreeListAdapter.js (hunk 17 above)
Comment (trickreich) [reply src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/adapters/TreeListAdapter.js]: Will be delivered

**Comment by danrot** [comment]
The build is failing, and there is still the empty `TreeRow.js` file.

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Body.js [hunk 26]
```diff
@@ -40,6 +54,18 @@ export default class Body extends React.PureComponent<Props> {
...
     };
 
+    handleRowExpand = (rowId: string | number) => {
+        if (this.props.onRowExpand) {
+            this.props.onRowExpand(rowId);
+        }
+    };
+
+    handleRowCollapse = (rowId: string | number) => {
+        if (this.props.onRowCollapse) {
```
[the remark this answers is not in the document]
Comment (danrot) [reply src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Body.js]: Seems like this one slipped through.

File: src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/tests/adapters/TreeListAdapter.test.js (hunk 20 above)
Comment (danrot) [reply src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/tests/adapters/TreeListAdapter.test.js]: I still don't see the difference between this test and `Render data with pencil button when onItemEdit callback is passed`. Except for the title being wrong :wink:

File: src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/adapters/TreeListAdapter.js [hunk 27]
```diff
@@ -0,0 +1,132 @@
...
+import FullLoadingStrategy from '../loadingStrategies/FullLoadingStrategy';
+import AbstractTableAdapter from './AbstractTableAdapter';
+
+@observer
+export default class TreeListAdapter extends AbstractTableAdapter {
+    static LoadingStrategy = FullLoadingStrategy;
+
+    static StructureStrategy = TreeStructureStrategy;
+
+    // TODO: When it's creating use the correct icon here
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/adapters/TreeListAdapter.js]: `When it's created`

**Comment by danrot** [comment]
- The column_list should be the default in the `WebspaceOverview`
- The id header has different margins than the title

**Comment by danrot** [comment]
- A click on the checkbox does nothing
